Getters for all the set socket options

avrecko opened the pull request:
Imho pretty obvious to have.

avrecko commented:
Build failed because test used a hardcoded port that was in use. Using an arbitrary free port instead. All good now.

apangin commented on src/one/nio/net/native/socket.c:
```diff
@@ -88,10 +88,19 @@ static int sockaddr_from_java(JNIEnv* env, jbyteArray address, jint port, struct
+    if(getsockopt(fd, level, optname, &optval, &length) == 0){
+        return optval;
+    }
+    throw_io_exception(env);
```
Throwing `IOException` from methods not declared to throw any unchecked exception.

avrecko replied:
I guess this is equivalent to JavaInternals#uncheckedThrow.

I wouldn't like to force users to check for exception while calling e.g. getRecvBuffer. I also wouldn't like to return 0 or false if getsockopt returned error. I think it makes sense to raise an exception.

What would you suggest?

apangin replied:
JavaInternals.uncheckedThrow is not encouraged either.

I don't expect getsockopt to fail unless it is called on a closed socket. For this reason I'm also against declaring `IOException` in method signatures. So, the question is what to return when calling getters on a closed socket. `getLocalAddress()`/`getRemoteAddress()` already return null in this case, and I think this is acceptable. So, I don't mind returning 0/false for other options, too. `IllegalStateException` also seems a reasonable outcome - indeed it means an operation on a socket in a wrong state. And, unlike `IOException`, it does not fool user by throwing unspecified exception.

avrecko replied:
Agree fully. Also thinking now. Failing with ISE doesn't add much added value as the code is going to fail on read/write anyway.

I've changed the code to be consistent with returning 0/false instead of throwing exception.

apangin commented on src/one/nio/net/native/socket.c:
```diff
@@ -520,34 +535,69 @@ Java_one_nio_net_NativeSocket_setTimeout(JNIEnv* env, jobject self, jint timeout
+    struct timeval tv;
+    socklen_t len = sizeof(tv);
+    if(getsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, &tv, &len)== 0){
+       return (tv.tv_sec * 1000) + (tv.tv_usec / 1000);
```
Isn't there a compiler warning about implicit casting `long` to `jint`?

avrecko replied:
I think the warning needs to be explicitly enabled with -Wconversion other-wize no warning.

The value can be set via the setTimeout(int timeout) imho it makes sense for getTimeout to return int and not long.

apangin commented on test/one/nio/http/HttpMethodTest.java:
```diff
@@ -34,15 +35,19 @@
+        int availablePort = dummySocket.getLocalPort();
+        dummySocket.close();
+        server = new TestServer(HttpServerConfigFactory.create(availablePort));
+        URL += ":" + availablePort;
```
There is no need to update URL as it is not used anywhere exception this method.

avrecko replied:
Good idea.

avrecko commented:
I've cleaned up some of the native code for getters.

In the http tests, sometimes the HttpClient requests happen faster than the AcceptorThread starts. In such cases there is Connection refused exception (in native impl there is no wait/retry).

I think it would be cool if the Server had a blocking start method (block until acceptors are up and running) and a way to query if acceptors are running. What do you think? Something like Server#startSync() and Server#isAcceptingRequests().

apangin commented:
I agree, it's better to just make `start` synchronous with the acceptor threads. I've committed the fix (along with some other small fixes from our internal repo).

apangin commented on test/one/nio/http/VirtualHostTest.java:
```diff
@@ -45,7 +48,17 @@ public static void beforeAll() throws IOException {
+        while (System.currentTimeMillis() < timeout && !exit) {
+            out.reset();
+            ThreadDumper.dump(out);
+            exit = out.toString().contains("NIO Acceptor");
```
Even if Acceptor has started, it does not yet mean it has called `listen` on the socket.
Checking the availability with a thread dump is not a pretty idea anyway - hopefully this is not needed anymore after the latest fix.

avrecko replied:
Totally agree, it was just a quick/ugly workaround. I would do start the same way with CDLatch. Good idea to have only sync version. Something like start(boolean sync) adds little value as the delay of sync version is going to be negligible anyway.

But I would on the other hand make start throw exception if not all acceptors can open a server socket. Maybe ssl port is already taken by a different process or something like that.

apangin replied:
I thought about making `start` throw `IOException`, but it's not obvious what to do when one of the acceptors fails while the other succeeds and starts serving requests.

apangin commented:
Thank you for addressing all the review comments - the PR has finally got merged.

avrecko commented:
It is a pleasure to work with this library. Thank you for making it
available on GitHub.

On Sun, 9 Jun 2019 at 14:25, Andrei Pangin <notifications@github.com> wrote:

> Thank you for addressing all the review comments - the PR has finally got
> merged.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/odnoklassniki/one-nio/pull/44?email_source=notifications&email_token=AACRDOLXR7D443WDPPEJKL3PZTZE3A5CNFSM4HQGGR2KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXIJDXQ#issuecomment-500208094>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACRDON4PETHB5U75QFQNVTPZTZE3ANCNFSM4HQGGR2A>
> .
>
